Faster Sparse Covariance Matrix

jeffwong opened the pull request:
This PR is for https://github.com/JuliaLang/LinearAlgebra.jl/issues/447, making cov faster for sparse matrices.

I am new to Julia so will need some help to make sure I have done my first PR correctly. From my machine I see a ~50x speedup in calling `cov` on a sparse matrix.

```
n = 50000
p = 500
x_sparse = sprand(n, p, .05)
@time cov1 = cov(x_sparse)
@time cov2 = spcov(x_sparse)
```

The two techniques produce answers that only differ by 1e-16.

Fixes https://github.com/JuliaLang/julia/issues/7788.

jeffwong committed:
optimize sparse covariance matrix

jeffwong committed:
test for optimized sparse covariance matrix

fredrikekre commented on base/sparse/linalg.jl:
```diff
@@ -941,3 +941,26 @@ end
 lu(A::SparseMatrixCSC) = error("Use lufact() instead of lu() for sparse matrices.")
 eig(A::SparseMatrixCSC) = error("Use eigs() instead of eig() for sparse matrices.")
+
+function spcov(X::SparseMatrixCSC{Tv1,Ti1} where {Tv1,Ti1})
```
`cov` instead of `spcov`?

ararslan replied:
Yes, as Fredrik points out, this should be `cov` since we want it to be a method of `Base.cov`. That way if someone calls `cov` on a matrix that happens to be sparse, they get the better-performing version without having to remember to call `spcov`.

nalimilan replied:
Also the type parameters are not used.

fredrikekre commented:
Btw, checkout the [BenchmarkTools.jl](https://github.com/JuliaCI/BenchmarkTools.jl) package for proper benchmarking :)

ararslan commented on test/sparse/sparse.jl:
```diff
@@ -1843,3 +1843,13 @@ end
+    x_dense = convert(Matrix{Float64}, x_sparse)
+    cov1 = cov(x_dense)
+    cov2 = spcov(x_sparse)
+    @test cov1 \approx cov2
```
Looks like you forgot to hit <kbd>tab</kbd> after `\approx` to get `≈`

ararslan commented on base/sparse/linalg.jl:
```diff
@@ -941,3 +941,26 @@ end
+
+    #Part 3 compute μ'μ
+    part3 = zeros(p,p)
+    for i in 1:p
```
You might be able to get a marginal speed boost by adding an `@inbounds` annotation, which avoids bounds checking in `getindex` and the like. For example, this could be written as
```julia
@inbounds for i = 1:p, j = i:p
    part3[i,j] = part3[j,i] = means[i] * means[j] * n
end
```
The simultaneous assignment as I've done above may or may not help, I actually don't know.

fredrikekre replied:
FWIW, I did some benchmarking, and we might as well just loop through the full matrix:
```julia
@inbounds for j = 1:p, i = 1:p
    part3[i,j] = means[i] * means[j] * n
end
```
Note the flipped index order though, since Julia is column major.

fredrikekre commented:
The problem is here: https://github.com/JuliaLang/julia/blob/1e95e1d9b6fbdba1840f0bb91c4c7acf1683bc62/base/statistics.jl#L332-L333 where we obtain a dense matrix from the `-` call. I suggest we add a method of `covm` specialized for sparse matrices instead. Then the entry point to `cov` stays the same and thus the keyword arg `corrected` works and it will also be possible to supply a dimension, ref docs for `cov(::AbstracMatrix)`:

```
    cov(X::AbstractMatrix[, vardim::Int=1]; corrected::Bool=true)

Compute the covariance matrix of the matrix X along the dimension vardim. If corrected is true (the default) then the sum is scaled with n-1, whereas the sum is scaled with n if corrected is false where n =
size(X, vardim).
```

nalimilan commented on test/sparse/sparse.jl:
```diff
@@ -1843,3 +1843,13 @@ end
 end
+
+@testset "optimizing sparse covariance" begin
+    n = 10000
```
Better use a smaller number to limit the memory use when running tests. Should also call `srand(1)` to similar to ensure results are reproducible.

nalimilan commented on base/sparse/linalg.jl:
```diff
@@ -941,3 +941,26 @@ end
+        end
+    end
+
+    return (part1 - part2 - part2' + part3) ./ (n-1)
```
Looking at this, I wonder whether you could avoid allocating temporary matrices for some of `part1`, `part2` and `part3`. For example, would it be possible to directly allocate the resulting matrix, and fill it element by element, at least for some tests? AFAICT this should be possible for `part3` at least.

jeffwong replied:
There probably is a way. For part1 (X'X) it is equivalent to doing dot(X[:,i], X[:,j]). But I wonder if we will lose out by not using the BLAS call for X'X? Also, if we leave it like this the weighted version that would go into JuliaStats is simple to extend (for example X' W X is the analogy for part1)

jeffwong replied:
Ah actually @fredrikekre already worked it out

fredrikekre commented:
I got carried away... Here is a modified version that includes the `vardim` and `corrected` kw, feel free to do whatever with it

```julia
function Base.cov(X::SparseMatrixCSC, vardim::Int=1; corrected::Bool=true)
    a, b = size(X)
    n, p = vardim == 1 ? (a, b) : (b, a)
    out = Matrix(Base.unscaled_covzm(X,vardim)) # part1
    sums = sum(X, vardim)
    means = mean(X, vardim)
    for j in 1:p, i in 1:p
        part2 = sums[i]*means[j]
        part3 = means[i] * means[j] * n
        out[i,j] += - part2 - conj(part2) + part3
    end
    return scale!(out, inv(n-Int(corrected)))
end
```

It allocates a bit less, but it is not much faster for the test case in the OP.

quinnj commented:
Great first contribution! 50x speedup is excellent!

ararslan commented:
Yes, this is indeed an excellent contribution! And don't be discouraged by the number of comments; everyone is just trying to pitch in to ensure that everything is as awesome as it can be. 🙂 Thanks for your hard work here!

jeffwong committed:
make spcov a function for the base method cov

jeffwong committed:
fix unit test

ararslan commented on base/sparse/linalg.jl:
```diff
@@ -941,3 +941,17 @@ end
+    out = Matrix(Base.unscaled_covzm(X,vardim)) # part1
+    sums = sum(X, vardim)
+    means = mean(X, vardim)
+    for j in 1:p, i in 1:p
```
Seems like an `@inbounds` annotation might still be useful here.

jeffwong commented:
The Travis-CI is taking an awfully long time (~ 4 hours)

I am seeing a nonzero exit code when building for [x86_64](https://travis-ci.org/JuliaLang/julia/jobs/252235714). However, it seems to have nothing to do with the unit tests... in fact it seems it cannot find julia itself? `/tmp/julia/bin/julia: No such file or directory`

```
The command "/tmp/julia/bin/julia --precompiled=no -e 'true' && /tmp/julia/bin/julia-debug --precompiled=no -e 'true'" exited with 127.
0.03s$ /tmp/julia/bin/julia -e 'versioninfo()'
/home/travis/.travis/job_stages: line 54: /tmp/julia/bin/julia: No such file or directory
The command "/tmp/julia/bin/julia -e 'versioninfo()'" exited with 127.
0.01s$ pushd /tmp/julia/share/julia/test
/home/travis/.travis/job_stages: line 54: pushd: /tmp/julia/share/julia/test: No such file or directory
The command "pushd /tmp/julia/share/julia/test" exited with 1.
0.01s$ export JULIA_CPU_CORES=2 && export JULIA_TEST_MAXRSS_MB=600 && /tmp/julia/bin/julia --check-bounds=yes runtests.jl $TESTSTORUN && /tmp/julia/bin/julia --check-bounds=yes runtests.jl libgit2-online download pkg
/home/travis/.travis/job_stages: line 54: /tmp/julia/bin/julia: No such file or directory
The command "export JULIA_CPU_CORES=2 && export JULIA_TEST_MAXRSS_MB=600 && /tmp/julia/bin/julia --check-bounds=yes runtests.jl $TESTSTORUN && /tmp/julia/bin/julia --check-bounds=yes runtests.jl libgit2-online download pkg" exited with 127.
0.02s$ popd
/home/travis/.travis/job_stages: line 54: popd: directory stack empty
The command "popd" exited with 1.
0.28s$ mkdir /tmp/embedding-test && make check -C /tmp/julia/share/doc/julia/examples/embedding \ JULIA="DYLD_FALLBACK_LIBRARY_PATH='$DYLD_FALLBACK_LIBRARY_PATH' /tmp/julia/bin/julia" \ BIN=/tmp/embedding-test \ "$(cd julia2 && make print-CC)"
make: *** /tmp/julia/share/doc/julia/examples/embedding: No such file or directory.  Stop.
The command "mkdir /tmp/embedding-test && make check -C /tmp/julia/share/doc/julia/examples/embedding \ JULIA="DYLD_FALLBACK_LIBRARY_PATH='$DYLD_FALLBACK_LIBRARY_PATH' /tmp/julia/bin/julia" \ BIN=/tmp/embedding-test \ "$(cd julia2 && make print-CC)"" exited with 2.
0.03s$ mv julia2 julia && rm -f julia/deps/scratch/libgit2-*/CMakeFiles/CMakeOutput.log
```

jeffwong committed:
use inbounds on for loop

fredrikekre commented on test/sparse/sparse.jl:
```diff
@@ -1843,3 +1843,14 @@ end
 end
+
+@testset "optimizing sparse covariance" begin
+    n = 1000
```
This can probably be even smaller, like 10 or something.

fredrikekre commented on test/sparse/sparse.jl:
```diff
@@ -1843,3 +1843,14 @@ end
+    x_dense = convert(Matrix{Float64}, x_sparse)
+    cov1 = cov(x_dense)
+    cov2 = cov(x_sparse)
+    @test cov1 ≈ cov2
```
We should probably also test the other input arguments. Suggestion:

```julia
@test cov(x_sparse)                     ≈ cov(x_dense)
@test cov(x_sparse, corrected=false)    ≈ cov(x_dense, corrected=false)
@test cov(x_sparse, 2)                  ≈ cov(x_dense, 2)
@test cov(x_sparse, 2, corrected=false) ≈ cov(x_dense, 2, corrected=false)
```

nalimilan replied:
It would also be good to test a matrix with `NaN` and another one with `Inf`, to ensure that the result is consistent with the standard method.

nalimilan commented on base/sparse/linalg.jl:
```diff
@@ -941,3 +941,17 @@ end
+function Base.cov(X::SparseMatrixCSC, vardim::Int=1; corrected::Bool=true)
+    a, b = size(X)
+    n, p = vardim == 1 ? (a, b) : (b, a)
+    out = Matrix(Base.unscaled_covzm(X,vardim)) # part1
```
Space after comma. Also `part1` cannot be understood by reading a code. Maybe add a comment before the function showing the formula with its three parts?

nalimilan commented:
The Travis failure was indeed unrelated, I've restarted the build.

andreasnoack commented on base/sparse/linalg.jl:
```diff
@@ -941,3 +941,17 @@ end
+    means = mean(X, vardim)
+    @inbounds for j in 1:p, i in 1:p
+        part2 = sums[i]*means[j]
+        part3 = means[i] * means[j] * n
```
Notice that `part2 == part3` in exact arithmetic so we can delete either `part2` or `part3`. I think `means[j]` should be `conj`ed here and the `conj` below shouldn't be there.

jeffwong replied:
Interesting catch! I will look into it after work today

jeffwong commented:
@fredrikekre it seems the for loop that computes out[i,j] could be parallelized. What are the best practices in Julia for parallelizing a for loop? Should base functions default to single threaded? Should I allow the number of threads to be a variable that is passed in?

I feel sparse cov should be parallelized as I imagine dense cov, which is just doing (X-mean)'(X-mean), is probably a parallelized matrix multiplication

KristofferC commented:
For now, threading in Julia is still a bit experimental so I would suggest not worrying about it for this PR.

jeffwong committed:
better comments and even faster implementation

jeffwong committed:
test for NaN and Inf

jeffwong commented:
@nalimilan It was good to suggest testing with NaN and Inf. I included it in the tests, and they are going to fail. When the matrix does not contain NaN/Inf the tests will pass. I actually don't know theoretically what a covariance matrix should be when there are NaN/Inf, can you help me understand what the dense one is doing?

KristofferC commented:
Oops, I approved under the assumption that CI would pass, I guess there is some discussion left what to do in the NaN Inf case.

Sacha0 commented on test/sparse/sparse.jl:
```diff
@@ -1843,3 +1843,39 @@ end
+    @test cov(x_sparse, 1, true)  ≈ cov(x_dense, 1, true)
+    @test cov(x_sparse, 1, false) ≈ cov(x_dense, 1, false)
+    @test cov(x_sparse, 2, true)  ≈ cov(x_dense, 2, true)
+    @test cov(x_sparse, 2, false) ≈ cov(x_dense, 2, false)
```
Perhaps DRY out these tests? For example, something along the lines of either of these sketches
```julia
@testset "optimizing sparse covariance" begin
    srand(1)
    x_sparse = sprand(10, 5, 0.5)
    x_dense = Matrix(x_sparse)
    for case in ("orig", "with NaN", "with Inf", "NaN & Inf")
        case == "with NaN" && (x_sparse[1] = NaN; x_dense[1] = NaN)
        case == "with Inf" && (x_sparse[1] = Inf; x_dense[1] = Inf)
        case == "NaN & Inf" && (x_sparse[[1, 2]] = [NaN, Inf]; x_dense[[1, 2]] = [NaN, Inf])
        @test cov(x_sparse, 1, true)  ≈ cov(x_dense, 1, true)
        @test cov(x_sparse, 1, false) ≈ cov(x_dense, 1, false)
        @test cov(x_sparse, 2, true)  ≈ cov(x_dense, 2, true)
        @test cov(x_sparse, 2, false) ≈ cov(x_dense, 2, false)
end

@testset "optimizing sparse covariance" begin
    srand(1)
    x_sparse = sprand(10, 5, 0.5)
    x_dense = Matrix(x_sparse)
    for (firstentry, secondentry) in (
            (x_sparse[1], x_sparse[2]), # original matrices
            (NaN,         x_sparse[2]), # with first entry NaN
            (Inf,         x_sparse[2]), # with first entry Inf
            (NaN,         Inf) )        # with first entry NaN, second Inf
        x_sparse[[1, 2]] = [firstentry, secondentry]
        x_dense[[1, 2]] = [firstentry, secondentry]
        @test cov(x_sparse, 1, true)  ≈ cov(x_dense, 1, true)
        @test cov(x_sparse, 1, false) ≈ cov(x_dense, 1, false)
        @test cov(x_sparse, 2, true)  ≈ cov(x_dense, 2, true)
        @test cov(x_sparse, 2, false) ≈ cov(x_dense, 2, false)
end
```
might do the trick? :)

nalimilan commented:
> @nalimilan It was good to suggest testing with NaN and Inf. I included it in the tests, and they are going to fail. When the matrix does not contain NaN/Inf the tests will pass. I actually don't know theoretically what a covariance matrix should be when there are NaN/Inf, can you help me understand what the dense one is doing?

I'm not familiar with that code, but AFAICT it simply propagates `NaN` and `Inf` so that any computation which involves such a value ends up `NaN`. This is just a consequence of the computation rules with these quantities: `NaN` gives `NaN` when combined with any number via arithmetic operations, and `Inf` gives `Inf` in most cases, except for `Inf/Inf`, which is `NaN`.

But the failure doesn't seem to indicate a problem in your algorithm, it's just that:
- you can't use `≈` or `==` to test for `NaN` or `Inf`: use `isequal` for that (and hardcode expected values if needed)
- you shouldn't call a deprecated function, which is an error when running CI: use the `corrected=true` or `corrected=false` keyword argument instead of the positional argument.

jeffwong commented:
I tried to update using the `corrected` keyword but got this error. I'm confused since corrected is a real keyword that is found [here](https://github.com/JuliaLang/julia/blob/master/base/statistics.jl#L379), but the error says it's not a keyword
```
julia> cov(x_dense,1,corrected=true)
ERROR: MethodError: no method matching cov(::Array{Float64,2}, ::Int64; corrected=true)
Closest candidates are:
  cov(::AbstractArray{T,2} where T, ::Int64) at statistics.jl:357 got unsupported keyword argument "corrected"
  cov(::AbstractArray{T,2} where T, ::Int64, ::Bool) at statistics.jl:357 got unsupported keyword argument "corrected"
  cov(::DenseArray{T,2} where T, ::StatsBase.AbstractWeights) at /Users/jeffrey/.julia/v0.6/StatsBase/src/cov.jl:96 got unsupported keyword argument "corrected"
  ...
```

jeffwong committed:
robust testing for NaN, Inf

jeffwong committed:
NaN and Inf testing

jeffwong commented:
The corrections to the tests using `isequal` for the NaN and Inf case seem to make things better. If we can resolve the issue here https://github.com/JuliaLang/julia/issues/22921 the tests should completely pass and I believe the PR will be done. I'll put this to the side and try to work on the `unscaled_covzm` issue

KristofferC commented:
Since the difference here seems to stem from a difference in how sparse matrix multiplication deals with Nans, Infs  (https://github.com/JuliaLang/julia/issues/22921#issuecomment-317238396) I am of the opinion that this PR should be merged. There is no need to block this PR on the eventual resolution of https://github.com/JuliaLang/julia/issues/22733.

nalimilan commented:
> I tried to update using the corrected keyword but got this error. I'm confused since corrected is a real keyword that is found here, but the error says it's not a keyword

Did you try this with a recent git master? These functions have been changed only a few weeks ago.

jeffwong committed:
fix corrected keyword

jeffwong commented:
I've messed up my git...I will fix soon

jeffwong commented:
You're right I needed to run make and run julia from source. I've updated again to use the corrected keyword. Now everything should be done! @KristofferC

nalimilan commented:
Thanks! Looks mostly good, but for the failing tests (with `Inf`) you should hardcode the values that the function currently returns (or replace `-Inf` with `NaN` in `cov_sparse`), put the `cov_dense` values in comments with a short explanation and a link to this PR and to those mentioned at https://github.com/JuliaLang/julia/pull/22735#issuecomment-317241324 above.

nalimilan commented on base/sparse/linalg.jl:
```diff
@@ -941,3 +941,37 @@ end
+
+    # Part1
+    # Compute X'X using sparse matrix operations
+    out = Matrix(Base.unscaled_covzm(X,vardim))
```
Missing space after comma.

StefanKarpinski commented:
I just wanted to say that this is a lovely PR.

jeffwong committed:
documentation about Inf

jeffwong committed:
style

jeffwong commented:
Update again @nalimilan @KristofferC. That should be the last of it

andreasnoack commented on base/sparse/linalg.jl:
```diff
@@ -941,3 +941,37 @@ end
+    a, b = size(X)
+    n, p = vardim == 1 ? (a, b) : (b, a)
+
+    # Cov(X) = E[(X-μ)'(X-μ)]
```
Maybe extend to `# Cov(X) = E[(X-μ)'(X-μ)] = E(XX') - μμ'` and simplify the comments below

andreasnoack replied:
Just pinging this comment because I think the PR is ready except for these comments which refer to the original version and therefore a bit misleading now. E.g. "Part 2" is no longer computed. I think that with the extra expression suggested in my previous comment, the comments in line 950-953 and 957-970 are redundant.

andreasnoack commented on base/sparse/linalg.jl:
```diff
@@ -941,3 +941,37 @@ end
+    #         [k1*k2*n k2*k2*n]
+    # but k1*n = a+b, and k2*n = c+d
+    sums = sum(X, vardim)
+    means = sums ./ n
```
If you move the `/n` into the loop below you'll avoid allocating this array.

jeffwong replied:
good catch

Sacha0 commented on test/sparse/sparse.jl:
```diff
@@ -1844,6 +1844,15 @@ end
+# of Inf in sparse matrix algebra
+# https://github.com/JuliaLang/julia/issues/22921
+# The issue will be resolved in
+# https://github.com/JuliaLang/julia/issues/22733
```
:100: for the superlative comments! :)

jeffwong committed:
less allocations

jeffwong commented:
I think travis CI failed for an unrelated reason, how can I restart it?

ararslan commented:
I agree the failures look unrelated. I restarted the failed builds anyway just in case. In general it requires commit access to restart specific builds, but if you want to restart all builds then you can simply close then immediately reopen the pull request from the GitHub UI. But since the builds take so long, it's better to be more conservative with which are restarted whenever possible.

nalimilan commented on base/sparse/linalg.jl:
```diff
@@ -968,9 +969,8 @@ function Base.cov(X::SparseMatrixCSC, vardim::Int=1; corrected::Bool=true)
-    means = sums ./ n
     @inbounds for j in 1:p, i in 1:p
-        part2 = sums[i] * means[j]
+        part2 = sums[i] * sums[j] / n
```
Maybe parentheses would be useful around the `/` to prevent overflow on integer matrices? (If multiplication is done first, overflow is much more likely, and computation will use the integer type rather than floating point.)

jeffwong committed:
parens to prevent overflow

jeffwong committed:
simplify comments

nalimilan commented:
Thanks @jeffwong! Looking forward to your future PRs.
